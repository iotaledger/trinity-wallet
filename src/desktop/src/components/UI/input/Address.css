.address-input {
    display: flex;
    justify-content: space-between;
    width: 100%;
    max-width: 512px;
    margin: 0 auto 30px;
    > div {
        display: block;
        position: relative;
        width: calc(100% - 70px);
    }
    input[type='text'] {
        font-family: 'Lato';
        font-weight: 200;
        font-size: var(--font-size-default);
        line-height: 26px;
        width: 100%;
        border: none;
        background: none;
        color: #fff;
        padding: 4px 0;
        font-size: 18px;
        resize: none;
        &:focus {
            outline: none;
        }
    }
    button {
        padding: 0 8px;
        height: 32px;
        min-width: 0px;
        font-size: 11px;
        margin: 1px 0 0;
        line-height: 18px;
        img {
            width: 14px;
            vertical-align: middle;
            margin: -2px 4px 0 0;
        }
    }
    small {
        position: absolute;
        bottom: 1px;
        left: 0px;
        display: block;
        width: 100%;
        text-align: left;
        &:before,
        &:after {
            position: absolute;
            content: '';
            width: 100%;
            height: 1px;
            background: white;
            top: 0px;
            left: 0px;
        }
        &:after {
            max-width: 0px;
            background: var(--color-yellow);
            transition: all 0.2s ease-out;
        }
    }
    input[type='text']:focus + small:after {
        max-width: 100%;
    }
}

.qrScanner {
    width: 640px;
    max-width: 100%;
    margin: auto;
    text-align: center;
    video {
        transform: rotateY(180deg);
        max-width: 100%;
        margin-bottom: 50px;
        border: 2px solid #fff;
    }
}
