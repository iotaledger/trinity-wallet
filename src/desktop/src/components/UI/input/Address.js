import React from 'react';
import QrReader from 'react-qr-reader';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { ADDRESS_LENGTH } from 'libs/util';
import css from './Input.css';
import Modal from 'components/UI/Modal';

import Camera from 'images/camera.png';

export default class AddressInput extends React.PureComponent {
    static propTypes = {
        address: PropTypes.string.isRequired,
        label: PropTypes.string.isRequired,
        closeLabel: PropTypes.string.isRequired,
    };

    state = {
        showScanner: false,
    };

    openScanner = e => {
        e.preventDefault();
        this.setState(() => ({
            showScanner: true,
        }));
    };

    closeScanner = e => {
        e.preventDefault();
        this.setState(() => ({
            showScanner: false,
        }));
    };

    onScanEvent = address => {
        if (address !== null) {
            this.setState(() => ({
                showScanner: false,
            }));
            this.props.onChange(address);
        }
    };

    onScanError = err => {
        console.log(err);
    };

    render() {
        const { address, label, closeLabel } = this.props;
        const { showScanner } = this.state;

        return (
            <div className={css.input}>
                <fieldset>
                    <a onClick={this.openScanner}>
                        <img src={Camera} alt="" />
                    </a>
                    <input
                        type="text"
                        value={address}
                        onChange={e => this.props.onChange(e.target.value)}
                        maxLength={ADDRESS_LENGTH}
                    />
                    <small>{label}</small>
                </fieldset>
                {showScanner && (
                    <Modal isOpen={true} onStateChange={showScanner => this.setState({ showScanner })} hideCloseButton>
                        <div className={css.qrScanner}>
                            <QrReader delay={350} onError={this.onScanError} onScan={this.onScanEvent} />
                            <Button type="button" onClick={this.closeScanner} variant="cta">
                                {closeLabel}
                            </Button>
                        </div>
                    </Modal>
                )}
            </div>
        );
    }
}
