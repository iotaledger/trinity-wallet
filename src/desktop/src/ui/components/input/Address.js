import React from 'react';
import QrReader from 'react-qr-reader';
import PropTypes from 'prop-types';
import { ADDRESS_LENGTH } from 'libs/util';

import Modal from 'ui/components/modal/Modal';
import Button from 'ui/components/Button';
import css from './input.css';

import Camera from 'images/camera-white.png';

/**
 * Address input component
 */
export default class AddressInput extends React.PureComponent {
    static propTypes = {
        /* Current address value */
        address: PropTypes.string.isRequired,
        /* Address input label */
        label: PropTypes.string.isRequired,
        /* Camera modal close button label */
        closeLabel: PropTypes.string.isRequired,
        /* Address change event function
         * @param {string} value - Current address value
        */
        onChange: PropTypes.func.isRequired,
    };

    state = {
        showScanner: false,
    };

    onScanEvent = (address) => {
        if (address !== null) {
            this.setState(() => ({
                showScanner: false,
            }));
            this.props.onChange(address);
        }
    };

    closeScanner = () => {
        this.setState(() => ({
            showScanner: false,
        }));
    };

    openScanner = () => {
        this.setState(() => ({
            showScanner: true,
        }));
    };

    render() {
        const { address, label, closeLabel } = this.props;
        const { showScanner } = this.state;

        return (
            <div className={css.input}>
                <fieldset>
                    <a onClick={this.openScanner}>
                        <img src={Camera} alt="" />
                    </a>
                    <input
                        type="text"
                        value={address}
                        onChange={(e) => this.props.onChange(e.target.value)}
                        maxLength={ADDRESS_LENGTH}
                    />
                    <small>{label}</small>
                </fieldset>
                {showScanner && (
                    <Modal isOpen onClose={this.closeScanner}>
                        <div className={css.qrScanner}>
                            <QrReader delay={350} onScan={this.onScanEvent} />
                            <Button type="button" onClick={this.closeScanner} variant="primary">
                                {closeLabel}
                            </Button>
                        </div>
                    </Modal>
                )}
            </div>
        );
    }
}
