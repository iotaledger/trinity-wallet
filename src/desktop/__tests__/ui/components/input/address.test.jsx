import React from 'react';
import { shallow } from 'enzyme';

import Address from 'ui/components/input/Address';

const props = {
    address: 'ABC',
    label: 'Foo',
    closeLabel: 'Bar',
    onChange: jest.fn(),
    verifyCDAContent: jest.fn()
};

describe('Address component', () => {
    test('Render the component', () => {
        const wrapper = shallow(<Address {...props} />);

        expect(wrapper).toMatchSnapshot();
    });

    test('Address value', () => {
        const wrapper = shallow(<Address {...props} />);

        expect(wrapper.find('input').props().value).toEqual('ABC');
    });

    test('Address change callback', () => {
        const wrapper = shallow(<Address {...props} />);

        wrapper.find('input').simulate('change', { target: { value: 'ABCDEF' } });
        expect(props.onChange).toHaveBeenLastCalledWith('ABCDEF');
    });

    test('Address label', () => {
        const wrapper = shallow(<Address {...props} />);

        expect(wrapper.find('small').text()).toEqual('Foo');
    });
});
