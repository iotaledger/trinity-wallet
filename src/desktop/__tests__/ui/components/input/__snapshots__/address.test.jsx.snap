// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Address component Render the component 1`] = `
<div
  className="input"
>
  <fieldset>
    <a
      onClick={[Function]}
    >
      <Icon
        icon="camera"
        size={16}
      />
    </a>
    <input
      data-tip="ABC"
      maxLength={90}
      onChange={[Function]}
      type="text"
      value="ABC"
    />
    <small>
      Foo
    </small>
    <p
      className="addressHidden"
    >
      ABC
    </p>
  </fieldset>
</div>
`;
